### ISSUE TITLE
terminationGracePeriodSeconds works with nanoserver based container images but not servercore based images

### ISSUE THREAD

**Issue by marosset** [microsoft/Windows-Containers#164]
Delayed app shutdown / terminationGracePeriod appear to only work with contianer images based on nanoserver.

I am experimenting with the `terminationGracePeriodSeconds' for Windows container applications and have observed that this functionally works as expected if the container images are based on nanoserver but do not behave as expected if the container images are based on servercore.

I have a kubernetes `v1.23.0-alpha.2` cluster with Windows Server 2019 nodes running containerd `v1.5.4`.

If I deploy the following pod to the cluster which runs this application https://github.com/marosset/windows-container-graceful-shutdown/blob/master/dotnet/Program.cs 

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: delayed-shutdown
  labels:
    app: delated-shutdown
spec:  
  replicas: 1
  template:
    metadata:
      name: delayed-shutdown
      labels:
        app: delayed-shutdown
    spec:
      containers:
      - name: delayed-shutdown
        image: mrosse3/delayed-shutdown-dotnet:5.0-nanoserver
        imagePullPolicy: Always
      terminationGracePeriodSeconds: 75
      nodeSelector:
        "kubernetes.io/os": windows
  selector:
    matchLabels:
      app: delayed-shutdown
```

If I follow the logs of one of the containers with `kubectl logs -f {pod name}` and then scale down the deployment or delete the pod I see

```cmd
Waiting for control event...
CTRL_SHUTDOWN received
Exiting in 60 seconds...
60
59
58
<snip>
4
3
2
1
Goodbye
```

If I set terminationGracePeriodSeconds to less than 60 seconds I see the log messages stop at the appropriate time which indicates that everything is working as expected.

If I run the same application in a container image based on mcr.microsoft.com/dotnet/runtime:5.0-windowsservercore-ltsc2019 (available at mrosse3/delayed-shutdown-dotnet:5.0-servercore) and view logs as the pod is being stopped I only see the following before the container exits.

```cmd
Waiting for control event...
CTRL_SHUTDOWN received 
```

When I try this in an application written in golang (https://github.com/marosset/windows-container-graceful-shutdown/blob/master/go/delayed-shutdown.go) I see the same behavior (except that on servercore the container waits 5 seconds before exiting regardless of what the specified terminationGracePeriodSeconds is set to)

**Comment by ghost** [comment]
This issue has been open for 30 days with no updates.
@judyliu-ms, please provide an update or close this issue.

**Comment by cwilhit** [comment]
Marking as in progress. MSFT internal 36689012 for reference.

**Comment by chrismilson** [comment]
We came across the same issue recently, and we could get the desired behaviour by setting the registry key, `HKLM:\SYSTEM\CurrentControlSet\Control\WaitToKillServiceTimeout`, to something higher than the `terminationGracePeriod`.

**Comment by acerock** [comment]
Any update on this? 
It is a critical feature for our workload where we have issues with auto-scaling and scale-in (pod termination). 

I can reproduce the different behavior running pods in AKS 1.23.8 windows nodes.
I run same app logic on both mcr.microsoft.com/dotnet/aspnet:6.0-nanoserver-1809 and mcr.microsoft.com/dotnet/framework/runtime:4.8-windowsservercore-ltsc2019. 

TerminationGracePeriodSeconds is honored for nanoserver but not for windows server core.

Adding the registry settings for windows server core don't help: 
```
USER ContainerAdministrator
RUN reg add hklm\system\currentcontrolset\services\cexecsvc /v ProcessShutdownTimeoutSeconds /t REG_DWORD /d 300  
RUN reg add hklm\system\currentcontrolset\control /v WaitToKillServiceTimeout /t REG_SZ /d 300000 /f
```

I'm pretty sure this used to work at some point, so I'm surprised we now see issues in prod environments.

Update:
Made workaround work by setting WaitToKillServiceTimeout (cexecsvc reg key not needed) to hardcoded max.
Then, just as important, the handler code receiving the CTRL_SHUTDOWN_EVENT signal need to block until main() has completed and returned. Without blocking the container just dies after 5 seconds. And if returning before main() you will see exit code -1073741510 (0xc000013a) in the kube-audit-admin logs (DeleteOptions requsts).

None of this is needed for nano server pods, they just work.

**Comment by kiashok** [comment]
Fixed with PR https://github.com/microsoft/hcsshim/pull/1416

**Comment by marosset** [comment]
Should we keep this open until the hcsshim updates are available in a release of containerd?

**Comment by brasmith-ms** [comment]
Containerd 1.7 has hcsshim version 0.9.2. This change is included in hcsshim version 0.9.8. https://github.com/microsoft/hcsshim/releases/tag/v0.9.8

**Comment by kiashok** [comment]
The fix has already made it to 1.7 and 1.6; 1.7 is using hcsshim v0.10.0-rc.7 https://github.com/containerd/containerd/blob/release/1.7/go.mod#L9 . The graceful termination fix went into v0.10.0-rc.4  itself https://github.com/microsoft/hcsshim/releases/tag/v0.10.0-rc.4

**Comment by kiashok** [comment]
1.6 is using hcsshim tag v0.9.8 https://github.com/containerd/containerd/blob/release/1.6/go.mod#L8 . the fix made it to v0.9.7 https://github.com/microsoft/hcsshim/releases/tag/v0.9.7

**Comment by chamindac** [comment]
I had same issue and made a work around solution with pre stop hook for dotnet wondows app containers. For me issue is happening in mcr.microsoft.com/dotnet/runtime:8.0-windowsservercore-ltsc2022 and workaround solution tested with the same.
https://chamindac.blogspot.com/2025/03/gracefully-shut-down-dotnet-8.html
